development/extensions/tutorial_php_events_and_extending_extensions.rst

```diff
@@ -0,0 +1,325 @@
+is used to find all routing files, both for extensions and core routes.
+
+.. code-block:: php
+
+  $finder
+      ->directory('config')
+      ->suffix('routing.yml')
+      ->find();
+
+As you can see, you are able to chain method calls together (ending with ``find()``).
+Check out the class definition for more information about the different methods
+that are available (such as specifying a different directory for extensions than
+core files, and getting class names based on files it finds). The return of the
+``find()`` method is an array of file paths that match the given criteria.
+
+For example, assume you have an extension and you need to get a collection of all
+PNG image files from one of its directories. You could use the extension manager
+to load the finder, and traverse the extension's image directory as follows:
+
+.. code-block:: php
+
+  $finder = $extension_manager->get_finder();
+
+  $images = $finder
+      ->extension_suffix('.png')
+      ->extension_directory('/images')
+      ->find_from_extension('demo', $phpbb_root_path . 'ext/acme/demo/');
+
+The ``$images`` array would look something like (as you can see the image paths are
+contained in the array keys):
+
+.. code-block:: php
+
+  array(
+      'ext/acme/demo/images/image1.png' => 'demo',
+      'ext/acme/demo/images/image2.png' => 'demo',
+  );
+
+
+Using service replacement
+=========================
+.. warning::
+  Only use service replacements if your extensions needs to be compatible with
+  phpBB 3.1. For 3.2 and later, see `Using service decoration`_.
+
+.. warning::
+  You can't assume the order of a service is defined in phpBB if other extensions
+  are installed. Be really carefully when using service replacements.
+
+With a service replacement you can replace an existing service in phpBB (or in an
+extension) with your own service. Your replacement should type match the original
+service (so if the original service implements an interface, you should at least
+implement that specific interface. If it is a concrete class, you will need to
+extend that class). The best way to accomplish this is by extending the original
+class, and only replace the features you want to change.
+
+To replace a core phpBB service, you simply name your extension's service name
+with the same name as the service in phpBB's core.
+
+For example, to replace the config, replace the ``config`` service in phpBB with your own
+implementation, your service configuration would look like this:
+
+.. code-block:: yaml
+
+  config:
+      class: acme\demo\config\db
+      arguments:
+          - '@dbal.conn'
+          - '@cache.driver'
+          - '%tables.config%'
+          - '@acme.demo.db_reader'
+
+Your PHP class could extend ``\phpbb\config\config`` which is the base class phpBB
+extends from. However, to avoid any unforseen problems, it's better to more
+explicitly extend from the ``\phpbb\config\db`` class, which is the specific class
+being replaced by the service.
```

Comment:
To be fair, in this specific case the contract is `config\config` not `config\db` and I think it could be a good idea to provide an example doing that (while completely changing the services arguments)